### PR TITLE
Refactor of Microsoft.Bot.Builder.Azure.Queues.

### LINKED ISSUE #4424: Update Bot.Builder.Azure.Queues
- Create an interface for the storage.
- Pull dialog specific code into Adaptive Dialogs

[enhancement]

### PR DESCRIPTION
**Pull request by mrivera-ms**
Fixes #4424

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/ContinueConversationLater.cs [hunk 1]
```diff
@@ -32,17 +32,21 @@ public class ContinueConversationLater : Dialog
...
+
+        private readonly QueueStorage _queueStorage;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="ContinueConversationLater"/> class.
         /// </summary>
+        /// <param name="queueStorage">The queue to store the Activity to continue the conversation.</param>
         /// <param name="callerPath">The full path of the source file that contains this called.</param>
         /// <param name="callerLine">The line within the source file that contains this caller.</param>
         [JsonConstructor]
```
Comment (EricDahlvang) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/ContinueConversationLater.cs:45]: I think the JsonConstructor needs to be 
`public ContinueConversationLater([CallerFilePath] string callerPath = "", [CallerLineNumber] int callerLine = 0)`

This class should have a load method retrieving the QueueStorage from TurnState, similar to how BeginSkill loads the DialogOptions:

https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/BeginSkill.cs#L313

this.dialogManager.InitialTurnState.Set(queuStorage);

Comment (EricDahlvang) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/ContinueConversationLater.cs:45]: or take the config parameters, and create the QueueStorage when needed?  (but, TurnState approach seems to be the pre-existing pattern)

**Review by EricDahlvang — CHANGES_REQUESTED** [review]

File: libraries/Microsoft.Bot.Builder.Azure.Queues/AzureQueueStorage.cs [hunk 2]
```diff
@@ -44,25 +44,24 @@ public QueuesStorage(string queuesStorageConnectionString, string queueName, Jso
...
         /// </summary>
         /// <param name="activity">This is expected to be an <see cref="Activity"/> retrieved from a call to 
         /// activity.GetConversationReference().GetContinuationActivity().  This enables restarting the conversation
         /// using BotAdapter.ContinueConversationAsync.</param>
-        /// <param name="visibilityTimeout">Default value of 0. Cannot be larger than 7 days.</param>
         /// <param name="timeToLive">Specifies the time-to-live interval for the message.</param>
         /// <param name="cancellationToken">Cancellation token for the async operation.</param>
         /// <returns><see cref="SendReceipt"/> as a Json string, from the QueueClient SendMessageAsync operation.</returns>
-        public async Task<string> QueueActivityAsync(Activity activity, TimeSpan? visibilityTimeout = null, TimeSpan? timeToLive = null,  CancellationToken cancellationToken = default)
+        public override async Task<string> QueueActivityAsync(Activity activity, TimeSpan? timeToLive = null, CancellationToken cancellationToken = default)
```
[the remark this answers is not in the document]
Comment (mrivera-ms) [reply libraries/Microsoft.Bot.Builder.Azure.Queues/AzureQueueStorage.cs]: I was thinking the timeout does not add value to the developer if Peek is not implemented, but I see that it is used. Adding it back.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/ContinueConversationLater.cs (hunk 1 above)
Comment (mrivera-ms) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/ContinueConversationLater.cs]: I am removing QueueStorage from the constructor and getting it from TurnState.

**Review by EricDahlvang — APPROVED** [review]
lgtm 

@tomlm does this right to you?

**Comment by mrivera-ms** [comment]
@gabog , @chrimc62 , could you please review this change? Thank you.

File: libraries/Microsoft.Bot.Builder/QueueStorage.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
+﻿using System;
```
Comment (gabog) [inline libraries/Microsoft.Bot.Builder/QueueStorage.cs:1]: Missing (C) header

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/ContinueConversationLater.cs [hunk 4]
```diff
@@ -32,7 +31,7 @@ public class ContinueConversationLater : Dialog
         /// The Kind name for this dialog.
         /// </summary>
         [JsonProperty("$kind")]
-        public const string Kind = "AzureQueues.ContinueConversationLater";
+        public const string Kind = "Microsoft.ContinueConversationLater";
```
Comment (gabog) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/ContinueConversationLater.cs:34]: Not sure if we still do this, but I think we need to update test.schema with the new name (I don't know if test.schema gets regenerated on the build server but the copy in source control has the old name)

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/ContinueConversationLater.cs [hunk 5]
```diff
@@ -6,10 +6,9 @@
...
 using System.Threading.Tasks;
 using AdaptiveExpressions.Properties;
-using Microsoft.Bot.Builder.Dialogs;
 using Newtonsoft.Json;
 
-namespace Microsoft.Bot.Builder.Azure.Queues
+namespace Microsoft.Bot.Builder.Dialogs.Adaptive.Actions
 {
     /// <summary>
     /// Action which schedules a conversation to be continued later by writing a EventActivity(Name=ContinueConversation) to a Azure Storage queue.
```
Comment (gabog) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/ContinueConversationLater.cs:14]: The summary says "Azure Storage queue", I think that after your refactoring to a base class we can just say queue (an not mention Azure)

**Review by gabog — CHANGES_REQUESTED** [review]
Just a few minor changes described in the comments.
I really like the introduction of the platform independent QueueStorage class and I think that now that we did that, we can remove the dependency on declarative from  Microsoft.Bot.Builder.Azure.Queues:

image

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/ContinueConversationLater.cs (hunk 5 above)
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/ContinueConversationLater.cs:14]: >a [](start = 79, length = 1)

an

File: libraries/Microsoft.Bot.Builder.Azure.Queues/Schemas/AzureQueues.ContinueConversationLater.schema [hunk 6]
```diff
@@ -30,23 +30,6 @@
...
-            ]
-        },
-        "queueName": {
-            "$ref": "schema:#/definitions/stringExpression",
-            "title": "Queue Name",
-            "description": "Name of the queue that your azure function is bound to.",
-            "examples": [
-                "activities"
-            ],
-            "default": "activities"
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Azure.Queues/Schemas/AzureQueues.ContinueConversationLater.schema:48]: It looks like the class was changed to match the schema name is good, but if we have shipped this it is a breaking change.  Have we shipped it?

File: tests/Microsoft.Bot.Builder.Azure.Tests/AzureQueueTests.cs [hunk 7]
```diff
@@ -40,13 +41,15 @@ public async Task ContinueConversationLaterTests()
...
+                var queueStorage = new AzureQueueStorage(ConnectionString, queueName);
                 var dm = new DialogManager(new ContinueConversationLater()
                 {
-                    ConnectionString = ConnectionString,
-                    QueueName = queueName,
                     Date = "=addSeconds(utcNow(), 2)",
                     Value = "foo"
                 });
+
+                dm.InitialTurnState.Set<QueueStorage>(queueStorage);
```
Comment (chrimc62) [inline tests/Microsoft.Bot.Builder.Azure.Tests/AzureQueueTests.cs:51]: Are there any samples which did this in the old way?  Do we need a sample?

**Review by chrimc62 — CHANGES_REQUESTED** [review]
:clock1:

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/ContinueConversationLater.cs (hunk 4 above)
Comment (mrivera-ms) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/ContinueConversationLater.cs:34]: Do you mean the schema in libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Actions? If so, that change is included in this PR. I have not found any other references in the source. Please let me know.

---
In reply to: [501018849](https://github.com/microsoft/botbuilder-dotnet/pull/4692#discussion_r501018849) [](ancestors = 501018849)

File: libraries/Microsoft.Bot.Builder.Azure.Queues/Schemas/AzureQueues.ContinueConversationLater.schema (hunk 6 above)
Comment (mrivera-ms) [reply libraries/Microsoft.Bot.Builder.Azure.Queues/Schemas/AzureQueues.ContinueConversationLater.schema:48]: It shipped as a preview, so we should be safe to change it.

---
In reply to: [501232641](https://github.com/microsoft/botbuilder-dotnet/pull/4692#discussion_r501232641) [](ancestors = 501232641)

File: tests/Microsoft.Bot.Builder.Azure.Tests/AzureQueueTests.cs (hunk 7 above)
Comment (mrivera-ms) [reply tests/Microsoft.Bot.Builder.Azure.Tests/AzureQueueTests.cs:51]: There are no samples to update with this change. We do have a plan to add this to a sample later.

---
In reply to: [501233216](https://github.com/microsoft/botbuilder-dotnet/pull/4692#discussion_r501233216) [](ancestors = 501233216)

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/ContinueConversationLater.cs (hunk 4 above)
Comment (gabog) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/ContinueConversationLater.cs:34]: I searched on my local copy of schema.tests and I see a reference to the old name, I do see it too on your branch [here](https://github.com/microsoft/botbuilder-dotnet/blob/moniriv/queueref/tests/tests.schema#L8), [here](https://github.com/microsoft/botbuilder-dotnet/blob/moniriv/queueref/tests/tests.schema#L453) and in a couple of other places.

It seems that tests.schema gets regenerated by the unit tests so I am not sure if this is relevant or not (if tests.schema is regenerated every time we may want to added to the gitignore, would this be the case @chrimc62?)

**Review by gabog — APPROVED** [review]
My main issues have been addressed, the only one I am not sure about is the tests.schema one, maybe @chrimc62 can help clarify that, do we need that in git? or can we add it to gitignore (testui.schema seems to be also generated when I run the tests).

**Review by chrimc62 — APPROVED** [review]
:shipit:

**Pull request merged**